common/appointment/get.ts

```diff
@@ -0,0 +1,118 @@
+import { Lecture as Appointment } from '../entity/Lecture';
+import { getUserType, User } from '../user';
+import { prisma } from '../prisma';
+import { AttendanceStatus } from '../entity/AppointmentAttendee';
+
+export const getAppointmentsForUser = async (user: User, take?: number, skip?: number): Promise<Appointment[]> => {
+    const userType = getUserType(user);
+    let appointmentIds: number[] = [];
+    switch (userType) {
+        case 'student':
+            appointmentIds = await getStudentAppointmentIds(user.studentId);
+            break;
+        case 'pupil':
+            appointmentIds = await getPupilAppointmentIds(user.pupilId);
+            break;
+        case 'screener':
+            appointmentIds = await getScreenerAppointmentIds(user.screenerId);
+    }
+    if (appointmentIds.length === 0) {
+        return null;
+    }
+
+    appointmentIds = deduplicate(sortIds(appointmentIds));
+
+    const appointments = await getAppointmentsByIdList(appointmentIds, take, skip);
+
+    if (userType !== 'pupil') {
+        return appointments;
+    }
+
+    /*
+     * in the first iteration pupils are not added as appointment participants automatically
+     * that's why we need to get appointments linked to subcourse and it's participants
+     * if pupils decline, than an entry in appointment_participant_pupil is created
+     *  */
+    const subcourseAppointments = await getPupilSubcourseAppointments(user.pupilId, take, skip);
+    return [...subcourseAppointments, ...appointments].sort((a, b) => b.start.getTime() - a.start.getTime());
+};
+
+const getAppointmentsByIdList = async (appointmentIds: number[], take, skip): Promise<Appointment[]> => {
+    return (await prisma.lecture.findMany({
+        where: {
+            id: {
+                in: appointmentIds,
+            },
+            AND: [{ isCanceled: false }, { isCanceled: null }], //@TODO: probably null will not be needed after field changes to not nullable
+        },
+        orderBy: [{ start: 'desc' }],
+        take,
+        skip,
+    })) as unknown as Appointment[];
+};
+
+const mergeIdLists = (...listsOfIds: number[][]): number[] => listsOfIds.reduce((flat, current) => flat.concat(current), []);
+
+const deduplicate = (list: number[]) => [...new Set(list)];
+
+const sortIds = (list: number[]) => [...new Uint32Array([...list]).sort()];
```

Comment:
Why conversion to a Typed Array and back?

Reply:
it is apparently more efficient for using the sort on numbers. I am actually not sure if we need to sort by id as we sort the results by date